File: library/src/com/devspark/sidenavigation/SideNavigationView.java
package com.devspark.sidenavigation;

import java.util.ArrayList;

import org.xmlpull.v1.XmlPullParser;

import android.content.Context;
import android.content.res.XmlResourceParser;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;

/**
 * View of displaying side navigation.
 * 
 * @author e.shishkin
 * 
 */
public class SideNavigationView extends LinearLayout {
    private static final String LOG_TAG = SideNavigationView.class.getSimpleName();

    private LinearLayout navigationMenu;
    private ListView listView;
    private View outsideView;

    private ISideNavigationCallback callback;
    private ArrayList<SideNavigationItem> menuItems;
    private Mode mMode = Mode.LEFT;

    public static enum Mode {
        LEFT, RIGHT
    };

    /**
     * Constructor of {@link SideNavigationView}.
     * 
     * @param context
     */
    public SideNavigationView(Context context) {
        super(context);
        load();
    }

    /**
     * Constructor of {@link SideNavigationView}.
     * 
     * @param context
     * @param attrs
     */
    public SideNavigationView(Context context, AttributeSet attrs) {
        super(context, attrs);
        load();
    }

    /**
     * Loading of side navigation view.
     */
    private void load() {
        if (isInEditMode()) {
            return;
        }
        initView();
    }

    /**
     * Initialization layout of side menu.
     */
    private void initView() {
        removeAllViews();
        int sideNavigationRes;
        switch (mMode) {
            case LEFT:
                sideNavigationRes = R.layout.side_navigation_left;
                break;
            case RIGHT:
                sideNavigationRes = R.layout.side_navigation_right;
                break;

            default:
                sideNavigationRes = R.layout.side_navigation_left;
                break;
        }
        LayoutInflater.from(getContext()).inflate(sideNavigationRes, this, true);
        navigationMenu = (LinearLayout) findViewById(R.id.side_navigation_menu);
        listView = (ListView) findViewById(R.id.side_navigation_listview);
        outsideView = (View) findViewById(R.id.side_navigation_outside_view);
        outsideView.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                hideMenu();
            }
        });
        listView.setOnItemClickListener(new OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                if (callback != null) {
                    callback.onSideNavigationItemClick(menuItems.get(position).getId());
                }
                hideMenu();
            }
        });
    }

    /**
     * Setup of {@link ISideNavigationCallback} for callback of item click.
     * 
     * @param callback
     */
    public void setMenuClickCallback(ISideNavigationCallback callback) {
        this.callback = callback;
    }

    /**
     * Setup of side menu items.
     * 
     * @param menu - resource ID
     */
    public void setMenuItems(int menu) {
        parseXml(menu);
        if (menuItems != null && menuItems.size() > 0) {
            listView.setAdapter(new SideNavigationAdapter());
        }
    }

    /**
     * Setup sliding mode of side menu ({@code Mode.LEFT} or {@code Mode.RIGHT}). {@code Mode.LEFT} by default.
     * 
     * @param mode Sliding mode
     */
    public void setMode(Mode mode) {
        if (isShown()) {
            hideMenu();
        }
        mMode = mode;
        initView();
        // setup menu items
        if (menuItems != null && menuItems.size() > 0) {
            listView.setAdapter(new SideNavigationAdapter());
        }
    }

    /**
     * Getting current side menu mode ({@code Mode.LEFT} or {@code Mode.RIGHT}). {@code Mode.LEFT} by default.
     * 
     * @return side menu mode
     */
    public Mode getMode() {
        return mMode;
    }

    /**
	 * 
	 */
    public void setBackgroundResource(int resource) {
        listView.setBackgroundResource(resource);
    }

    /**
     * Show side navigation menu.
     */
    public void showMenu() {
        outsideView.setVisibility(View.VISIBLE);
        outsideView.startAnimation(AnimationUtils.loadAnimation(getContext(), R.anim.side_navigation_fade_in));
        // show navigation menu with animation
        int animRes;
        switch (mMode) {
            case LEFT:
                animRes = R.anim.side_navigation_in_from_left;
                break;
            case RIGHT:
                animRes = R.anim.side_navigation_in_from_right;
                break;

            default:
                animRes = R.anim.side_navigation_in_from_left;
                break;
        }
        navigationMenu.setVisibility(View.VISIBLE);
        navigationMenu.startAnimation(AnimationUtils.loadAnimation(getContext(), animRes));
    }

    /**
     * Hide side navigation menu.
     */
    public void hideMenu() {
        outsideView.setVisibility(View.GONE);
        outsideView.startAnimation(AnimationUtils.loadAnimation(getContext(), R.anim.side_navigation_fade_out));
        // hide navigation menu with animation
        int animRes;
        switch (mMode) {
            case LEFT:
                animRes = R.anim.side_navigation_out_to_left;
                break;
            case RIGHT:
                animRes = R.anim.side_navigation_out_to_right;
                break;

            default:
                animRes = R.anim.side_navigation_out_to_left;
                break;
        }
        navigationMenu.setVisibility(View.GONE);
        navigationMenu.startAnimation(AnimationUtils.loadAnimation(getContext(), animRes));
    }

    /**
     * Show/Hide side navigation menu depending on visibility.
     */
    public void toggleMenu() {
        if (isShown()) {
            hideMenu();
        } else {
            showMenu();
        }
    }

    @Override
    public boolean isShown() {
        return navigationMenu.isShown();
    }

    /**
     * Parse XML describe menu.
     * 
     * @param menu - resource ID
     */
    private void parseXml(int menu) {
        menuItems = new ArrayList<SideNavigationItem>();
        try {
            XmlResourceParser xrp = getResources().getXml(menu);
            xrp.next();
            int eventType = xrp.getEventType();
            while (eventType != XmlPullParser.END_DOCUMENT) {
                if (eventType == XmlPullParser.START_TAG) {
                    String elemName = xrp.getName();
                    if (elemName.equals("item")) {
                        String textId = xrp.getAttributeValue(
                                "http://schemas.android.com/apk/res/android",
                                "title");
                        String iconId = xrp.getAttributeValue(
                                "http://schemas.android.com/apk/res/android",
                                "icon");
                        String resId = xrp.getAttributeValue(
                                "http://schemas.android.com/apk/res/android",
                                "id");
                        SideNavigationItem item = new SideNavigationItem();
                        item.setId(Integer.valueOf(resId.replace("@", "")));
                        item.setText(resourceIdToString(textId));
                        if (iconId != null) {
                            try {
                                item.setIcon(Integer.valueOf(iconId.replace("@", "")));
                            } catch (NumberFormatException e) {
                                Log.d(LOG_TAG, "Item " + item.getId() + " not have icon");
                            }
                        }
                        menuItems.add(item);
                    }
                }
                eventType = xrp.next();
            }
        } catch (Exception e) {
            Log.w(LOG_TAG, e);
        }
    }

    /**
     * Convert resource ID to String.
     * 
     * @param text
     * @return
     */
    private String resourceIdToString(String resId) {
        if (!resId.contains("@")) {
            return resId;
        } else {
            String id = resId.replace("@", "");
            return getResources().getString(Integer.valueOf(id));
        }
    }

    private class SideNavigationAdapter extends BaseAdapter {
        private LayoutInflater inflater;

        public SideNavigationAdapter() {
            inflater = LayoutInflater.from(getContext());
        }

        @Override
        public int getCount() {
            return menuItems.size();
        }

        @Override
        public Object getItem(int position) {
            return null;
        }

        @Override
        public long getItemId(int position) {
            return 0;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            final ViewHolder holder;
            if (convertView == null) {
                convertView = inflater.inflate(R.layout.side_navigation_item, null);
                holder = new ViewHolder();
                holder.text = (TextView) convertView.findViewById(R.id.side_navigation_item_text);
                holder.icon = (ImageView) convertView.findViewById(R.id.side_navigation_item_icon);
                convertView.setTag(holder);
            } else {
                holder = (ViewHolder) convertView.getTag();
            }

            SideNavigationItem item = menuItems.get(position);
            holder.text.setText(menuItems.get(position).getText());
            if (item.getIcon() != SideNavigationItem.DEFAULT_ICON_VALUE) {
                holder.icon.setVisibility(View.VISIBLE);
                holder.icon.setImageResource(menuItems.get(position).getIcon());
            } else {
                holder.icon.setVisibility(View.GONE);
            }
            return convertView;
        }

        class ViewHolder {
            TextView text;
            ImageView icon;
        }

    }

}

File: sample.net/SideNavigation/SideNavigationView.cs
using System.Collections.Generic;
using System.Xml;
using Android.Views.Animations;
using Android.Widget;
using Android.Content;
using Android.Views;
using Java.Lang;
using Android.Util;
using sample.net;

/**
 * 
 * @author hram
 * 
 */
namespace com.devspark.sidenavigation
{
    public class SideNavigationView : LinearLayout
    {
        private static string LOG_TAG = "SideNavigationView";

	    private LinearLayout navigationMenu;
	    private ListView listView;
	    private View outsideView;
	    private ISideNavigationCallback callback;
	    private static List<SideNavigationItem> menuItems;

	    public SideNavigationView(Context context) : base(context)
	    {
		    load();
	    }

        public SideNavigationView(Context context, Android.Util.IAttributeSet attrs) : base(context, attrs)
        {
		    load();
	    }

	    private void load() {
		    if (IsInEditMode) {
			    return;
		    }
		    inflateLayout();
		    initUi();
	    }

	    private void inflateLayout() {
		    try {
			    LayoutInflater.From(Context).Inflate(Resource.Layout.side_navigation, this, true);
		    } catch (Exception e) {
			    Log.Warn(LOG_TAG, e);
		    }
	    }

	    private void initUi() {
		    navigationMenu = FindViewById<LinearLayout>(Resource.Id.side_navigation_menu);
		    listView = FindViewById<ListView>(Resource.Id.side_navigation_listview);
		    outsideView = FindViewById<View>(Resource.Id.side_navigation_outside_view);
		
		    outsideView.Click += delegate 
            {
				    hideMenu();
		    };
		
		    listView.ItemClick += delegate (object sender, AdapterView.ItemClickEventArgs args) 
            {	
				    if (callback != null) {
					    callback.onSideNavigationItemClick(menuItems[args.Position].id);
				    }
				    hideMenu();
		    };
	    }

	    public void setMenuClickCallback(ISideNavigationCallback callback) {
		    this.callback = callback;
	    }

        public void resetMenuClickCallback()
        {
            this.callback = null;
        }

	    public void setMenuItems(int menu) {
		    parseXml(menu);
		    if (menuItems != null && menuItems.Count > 0) {
			    listView.Adapter = new Adapter(Context);
		    }
	    }

	    public void setBackgroundResource(int resource) {
		    listView.SetBackgroundResource(resource);
	    }

	    public void showMenu() 
	    {
		    outsideView.Visibility = ViewStates.Visible;
            outsideView.StartAnimation(AnimationUtils.LoadAnimation(Context, Resource.Animation.side_navigation_fade_in));
            navigationMenu.Visibility = ViewStates.Visible;
            navigationMenu.StartAnimation(AnimationUtils.LoadAnimation(Context, Resource.Animation.side_navigation_in_from_left));
	    }

	    public void hideMenu() 
	    {
		    outsideView.Visibility = ViewStates.Gone;
            outsideView.StartAnimation(AnimationUtils.LoadAnimation(Context, Resource.Animation.side_navigation_fade_out));
	        navigationMenu.Visibility = ViewStates.Gone;
            navigationMenu.StartAnimation(AnimationUtils.LoadAnimation(Context, Resource.Animation.side_navigation_out_to_left));
	    }

	    public void toggleMenu() {
            if (outsideView.Visibility == ViewStates.Gone)
            {
			    showMenu();
		    } else {
			    hideMenu();
		    }
	    }

	    private void parseXml(int menu) {
		    menuItems = new List<SideNavigationItem>();
		    try {
                System.Xml.XmlReader reader = Resources.GetXml(menu);

                var sb = new System.Text.StringBuilder();

                while (reader.Read())
                    sb.AppendLine(reader.ReadOuterXml());

                XmlDocument document = new XmlDocument();
		        document.LoadXml(sb.ToString());

                foreach (XmlNode node in document.GetElementsByTagName("item"))
                {
                    string textId = node.Attributes["title", "http://schemas.android.com/apk/res/android"].Value;
                    string iconId = node.Attributes["icon", "http://schemas.android.com/apk/res/android"].Value;
                    string resId = node.Attributes["id", "http://schemas.android.com/apk/res/android"].Value;

                    SideNavigationItem item = new SideNavigationItem();
                    item.id = int.Parse(resId.Replace("@", ""));
                    item.text = resourceIdToString(textId);
                    item.icon = int.Parse(iconId.Replace("@", ""));
                    menuItems.Add(item);
                }
		    } catch (Exception e) {
			    Log.Warn(LOG_TAG, e);
		    }
	    }

	    private string resourceIdToString(string text) {
		    if (!text.Contains("@")) {
			    return text;
		    } else {
			    string id = text.Replace("@", "");
			    return Resources.GetString(int.Parse(id));
		    }
	    }

	    class SideNavigationItem {
	        public int id;
            public string text;
            public int icon;
	    }

        private class Adapter : Android.Widget.BaseAdapter
        {
		    private LayoutInflater inflater;

		    public Adapter(Context context) {
                inflater = LayoutInflater.From(context);
		    }

            public override int Count { get { return menuItems.Count; } }

            public override Java.Lang.Object GetItem(int position)
            {
                return null;
            }

            public override long GetItemId(int position)
            {
                return 0;
            }

		    public override View GetView(int position, View convertView, ViewGroup parent) 
            {
			    ViewHolder holder;
			    if (convertView == null) {
				    convertView = inflater.Inflate(Resource.Layout.side_navigation_item, null);
				    holder = new ViewHolder();
                    holder.image = convertView.FindViewById<ImageView>(Resource.Id.side_navigation_item_icon);
                    holder.text = convertView.FindViewById<TextView>(Resource.Id.side_navigation_item_text);
				    convertView.Tag = holder;
			    } else {
				    holder = (ViewHolder) convertView.Tag;
			    }
			    holder.image.SetImageResource(menuItems[position].icon);
                holder.text.Text = menuItems[position].text;
			    return convertView;
		    }

		    class ViewHolder : Java.Lang.Object {
                public TextView text;
                public ImageView image;
		    }

	    }

    }
}